core/jobs/batch_jobs/mailchimp_population_jobs.py

```diff
@@ -0,0 +1,188 @@
+            yield result.Err('User update failed for: %s' % failed_emails)
+
+
+class CombineTuples(beam.CombineFn):  # type: ignore[misc]
+    """CombineFn for combining all user email, status tuples."""
+
+    def create_accumulator(self) -> List:
+        """Base accumulator where the tuples are added."""
+        return []
+
+    def add_input(self, accumulator: List, model: Tuple) -> List:
+        """Append each tuple to the accumulator list."""
+        accumulator.append(model)
+        return accumulator
+
+    def merge_accumulators(self, accumulators: Iterable[List]) -> List:
+        """Merging accumulators is just combining both of them into a single
+        list.
+        """
+        return list(accumulators)
+
+    def extract_output(
+        self, accumulator: List
+    ) -> List:
+        """Output is the accumulator itself."""
+        return accumulator
+
+
+class MailchimpPopulateJob(base_jobs.JobBase):
+    """One-off job for populating the mailchimp db."""
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        # Pcollection that returns the relevant config property with batch
+        # index.
+        config_property = (
+            self.pipeline
+            | 'Get all config properties' >> ndb_io.GetModels(
+                config_models.ConfigPropertyModel.get_all())
+            | 'Get the batch_index_for_mailchimp property value' >> beam.Filter(
+                lambda model: model.id == 'batch_index_for_mailchimp')
+            | 'Convert model into a dict' >> beam.Map(
+                lambda model: (model.id, model.value)))
+        batch_index_dict = beam.pvalue.AsDict(config_property)
+
+        # PCollection with all user ids that have opted in for email
+        # newsletters.
+        relevant_user_ids = (
+            self.pipeline
+            | 'Get all UserEmailPreferencesModel' >> ndb_io.GetModels(
+                user_models.UserEmailPreferencesModel.get_all().order(
+                    'created_on'))
+            | 'Filter for site_updates == true' >> beam.Filter(
+                lambda model: model.site_updates is True)
+            | 'Extract user id as tuple' >> beam.Map(
+                lambda preferences_model: (preferences_model.id, True))
+        )
+
+        # PCollection of all user ids in oppia mapped to their emails.
+        relevant_user_emails = (
+            self.pipeline
+            | 'Get all user settings models' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all())
+            | 'Extract user email' >> beam.Map(
+                lambda settings_model: (settings_model.id, settings_model.email)
+            )
+        )
+
+        # Merge both PCollections to map the user email to the user ids along
+        # with a bool 'True' if they have opted in for newsletter.
+        mailchimp_results = (
+            (relevant_user_emails, relevant_user_ids)
+            | 'Group by user_id' >> beam.CoGroupByKey()
```

Comment:
Instead of doing this you can just transform `relevant_user_ids` into list using `AsList` and then filter according to it here.

Reply:
But how do I get the email of the user then? The user preferences model has no email data, so I have to combine it with the email data, right?

Reply:
Done (using `AsIter`)